### PR TITLE
xgboost: R Package Support

### PR DESCRIPTION
**Pull request by nviets**
Added option to build the R library, which allows for GPU support. The R package build is not compatible with the python or CLI build because the xgboost.so it produces is not full featured. This build works and can be loaded directly in R via .libPaths. However, it is not integrated with the rest of the R packaging system in nix. WIP.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by nviets** [comment]
@alexvorobiev, @trivialfis - is it feasible to overlay the R build here into the main R packaging system within nix? The cran version there is fine in general, but I didn't see a way to overlay the Cran build to add GPU support. This branch builds and runs.

**Comment by nviets** [comment]
@jbedo, @cfhammill - I'd like to add the GPU version of xgboost to rpackages, but xgboost has a custom builder. This PR uses that builder to create the R Package. The result is usable, if I add it to the libPath directly from the nix store, but I'm not sure how it can be added to the existing r-modules infrastructure which is tuned specifically for cran and bioc. Any ideas would be appreciated!

**Comment by trivialfis** [comment]
considering autotools is the primary build tool used for R packages, I'm not sure how to get CUDA support there without first splitting up the c++ core and the R binding. I'm not familiar with autotools, so doing that might take a while.

**Comment by jbedo** [comment]
Isn't this on CRAN and so already in rPackages? Not sure about GPU support and what's required for that though.

**Comment by nviets** [comment]
@jbedo - only the CPU version of xgboost is in CRAN, and I'd like to add support for the GPU version. The two ways that come to mind are to either have r-modules work with non-CRAN libraries like what is built here, or update the xgboost R library to source its binary (xgboost.so) from here.

@trivialfis - would it be possible to add a build option in your R package to use a system xgboost.so binary rather than building internally? Then we could just add pkgs.xgboost as an input to the rPackages.xgboost similar to how other packages source dependencies.

**Comment by trivialfis** [comment]
Technically it's possible, we just need to figure out how to pass this information through R CMD to autobools, potentially through a build time env. Then we can have autotools to perform library lookup, potentially through the use of pkgcomfig, which is supported by xgboost.

However, users of this package might get upset since such build will deviate from other packages on cran. There are many customization to use functions like logging, math functions including gamma, and most importantly the global random engine from the R runtime instead of libc or libstdc++.

It will take lots of work to convert these customization from build time to run time. We do build a GPU version of R package on our CI, but it's quite hacky. Any suggestion is welcomed!

**Comment by jbedo** [comment]
There's no need to have the R output from this derivation be part of rPackages, indeed as rPackages is generated from CRAN it would be best not to and to keep rPackages purely CRAN/BioC, allowing users a choice between CRAN official and this GPU accelerated version. If this derivation propagates `R_LIBS_SITE` as the generic CRAN builder does it can be used with `rWrapper`.

**Comment by nviets** [comment]
@jbedo - thanks for the tip on R_LIBS_SITE. Could you share a reference on how I can set up the propagation?

**Comment by nviets** [comment]
It's almost working, but xgboost's dependencies aren't getting picked up:

```
➜ nix-build -E "with import /home/nviets/repos/nixpkgs {}; let xgb = xgboost.override{ rSupport = true; doCheck = false;}; in rWrapper.override{ packages = [ xgb ]; }"
/nix/store/j5ax0yca844gmvfslf8556s87zs2w0cr-R-4.2.2-wrapper

➜ ./result/bin/R

R version 4.2.2 (2022-10-31) -- "Innocent and Trusting"
Copyright (C) 2022 The R Foundation for Statistical Computing
Platform: x86_64-pc-linux-gnu (64-bit)

R is free software and comes with ABSOLUTELY NO WARRANTY.
You are welcome to redistribute it under certain conditions.
Type 'license()' or 'licence()' for distribution details.

  Natural language support but running in an English locale

R is a collaborative project with many contributors.
Type 'contributors()' for more information and
'citation()' on how to cite R or R packages in publications.

Type 'demo()' for some demos, 'help()' for on-line help, or
'help.start()' for an HTML browser interface to help.
Type 'q()' to quit R.

> library(xgboost)
Error: package or namespace load failed for ‘xgboost’ in loadNamespace(j <- i[[1L]], c(lib.loc, .libPaths()), versionCheck = vI[[j]]):
 there is no package called ‘data.table’
```

What's missing besides the [export for R_LIBS_SITE](https://github.com/nviets/nixpkgs/blob/xgboostR/pkgs/development/libraries/xgboost/default.nix#L84)?

@jbedo

**Comment by jbedo** [comment]
Looking at the PR you're handling everything ok (sorry for delay, was on holidays), you just need to move the R dependencies to `propagatedBuildInputs` from `nativeBuildInputs`.

**Comment by nviets** [comment]
No luck. The build fails on the same complaint that data.table isn't available. data.table and the xgboost's other dependencies aren't showing up in R's libPaths, so I think they aren't getting added to R_LIBS_SITE.

In xgboost's case, the trouble isn't a big deal because it only has two or three R library requirements, but it would be nice to solve it more generally. Catboost has a similar custom R library build process outside of CRAN, and I was hoping to work on it next.

**Comment by jbedo** [comment]
There's a typo in `propagatedBuildInputs`, and you have to patch cmake to install in the right directory. The following edits resolved the issues for me:

```
diff --git a/pkgs/development/libraries/xgboost/default.nix b/pkgs/development/libraries/xgboost/default.nix
index 6c2d056a9dd..7a55f3d4c54 100644
--- a/pkgs/development/libraries/xgboost/default.nix
+++ b/pkgs/development/libraries/xgboost/default.nix
@@ -39,16 +39,12 @@ stdenv.mkDerivation rec {
     cudaPackages.autoAddOpenGLRunpathHook
   ] ++ lib.optionals rLibrary [
     R
-    rPackages.data_table
-    rPackages.jsonlite
-    rPackages.Matrix
   ];
 
   buildInputs = [ gtest ] ++ lib.optional cudaSupport cudaPackages.cudatoolkit
     ++ lib.optional ncclSupport cudaPackages.nccl;
 
-  propogatedBuildInputs = lib.optionals rLibrary [
-    R
+  propagatedBuildInputs = lib.optionals rLibrary [
     rPackages.data_table
     rPackages.jsonlite
     rPackages.Matrix
@@ -60,6 +56,11 @@ stdenv.mkDerivation rec {
     ++ lib.optionals ncclSupport [ "-DUSE_NCCL=ON" ]
     ++ lib.optionals rLibrary [ "-DR_LIB=ON" ];
 
+  preConfigure = ''
+    substituteInPlace cmake/RPackageInstall.cmake.in --replace "CMD INSTALL" "CMD INSTALL -l $out/library"
+    export R_LIBS_SITE="$R_LIBS_SITE''${R_LIBS_SITE:+:}$out/library"
+  '';
+
   # tests are expected to fail if rLibrary is enabled
   inherit doCheck;

**Comment by nviets** [comment]
Thanks for the corrections, @jbedo! It's finally building. I'll submit the PR once I confirm it works with cuda and aarch64-darwin systems.

**Comment by nviets** [comment]
The R library in this branch builds on linux with and without CUDA support on x86 linux, but not on aarch64-darwin due to an issue with FindLibR.cmake:

```
...
-- Performing Test SUPPORT_MSSE2 - Success
CMake Error at /nix/store/l36kwcsnba2fmfhjlas4vshrdav3jcwz-cmake-3.24.3/share/cmake-3.24/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find LibR (missing: LIBR_HOME LIBR_CORE_LIBRARY)
Call Stack (most recent call first):
  /nix/store/l36kwcsnba2fmfhjlas4vshrdav3jcwz-cmake-3.24.3/share/cmake-3.24/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  cmake/modules/FindLibR.cmake:183 (find_package_handle_standard_args)
  R-package/CMakeLists.txt:1 (find_package)
...
```

Since CUDA support on Linux is the point of this build outside CRAN, I think it's fine to proceed. I suspect there's a simple patch for FindLibR.cmake if anyone cares to try. @trivialfis

The R package version builds, runs, and handles library paths correctly, and the CLI version and shared library for python continue to work as expected. I'm squashing the commits and submitting a PR as is.

Thanks @jbedo

**Comment by trivialfis** [comment]
Unfortunately, I'm not familiar with the macos environment.

**Comment by nviets** [comment]
Thanks for the feedback @trivialfis and @jbedo. This branch solves the primary use case for enabling CUDA for xgboost on x86_64-linux, so I think is ready for to merge.

**Comment by nviets** [comment]
Added comments and an example of R package usage. Squash commits. Ready for merge.

File: pkgs/development/libraries/xgboost/default.nix [hunk 1]
```diff
@@ -1,69 +1,111 @@
...
+    # libxgboost.so, which isn't full featured for python and CLI
+    + lib.optionalString rLibrary ''
+      mkdir $out/library
+      export R_LIBS_SITE="$out/library:$R_LIBS_SITE''${R_LIBS_SITE:+:}"
+      make install -l $out/library
+    '' + ''
+      runHook postInstall
+    '';
+
+  # either the propagated build inputs aren't working or there aren't any. needs review
```
Comment (jbedo) [inline pkgs/development/libraries/xgboost/default.nix:99]: I think this comment isn't valid anymore?
```suggestion
```

File: pkgs/development/libraries/xgboost/default.nix [hunk 2]
```diff
@@ -1,69 +1,111 @@
...
+    [ "-DBUILD_WITH_CUDA_CUB=ON" ]
+    ++ lib.optionals ncclSupport [ "-DUSE_NCCL=ON" ]
+    ++ lib.optionals rLibrary [ "-DR_LIB=ON" ];
 
+  preConfigure = lib.optionals rLibrary ''
+    substituteInPlace cmake/RPackageInstall.cmake.in --replace "CMD INSTALL" "CMD INSTALL -l $out/library"
+    export R_LIBS_SITE="$R_LIBS_SITE''${R_LIBS_SITE:+:}$out/library"
+  '';
+
+  # tests are expected to fail if rLibrary is enabled
```
Comment (jbedo) [inline pkgs/development/libraries/xgboost/default.nix:66]: This comment is better next to the assertion, and would be good to expand it a bit and say why it's expected.

**Review by jbedo — CHANGES_REQUESTED** [review]

**Comment by nviets** [comment]
Thanks for the merge

**Pull request merged**